module/icp/algs/sha2/sha2.c

```diff
@@ -54,17 +56,220 @@ static void Encode64(uint8_t *, uint64_t *, size_t);
 
 /* userspace only supports the generic version */
 #if	defined(__amd64) && defined(_KERNEL)
-#define	SHA512Transform(ctx, in) SHA512TransformBlocks((ctx), (in), 1)
-#define	SHA256Transform(ctx, in) SHA256TransformBlocks((ctx), (in), 1)
 
-void SHA512TransformBlocks(SHA2_CTX *ctx, const void *in, size_t num);
-void SHA256TransformBlocks(SHA2_CTX *ctx, const void *in, size_t num);
+#include <sys/simd.h>
+
+typedef void (*sha256_block_f)(uint32_t *state, const void *in, size_t num);
+typedef void (*sha512_block_f)(uint64_t *state, const void *in, size_t num);
 
-#else
-static void SHA256Transform(SHA2_CTX *, const uint8_t *);
-static void SHA512Transform(SHA2_CTX *, const uint8_t *);
 #endif	/* __amd64 && _KERNEL */
 
+#if defined(_KERNEL)
+
+#define	BENCHMARK_BUFFER_SIZE (32 * 1024)
+
+// SHA-256
+
+static void sha256_alg_impl_benchmark(
+    const alg_impl_ops_t *ops, void *buffer, size_t buffer_n)
+{
+	SHA2_CTX sha2_ctx;
+	sha256_block_f sha256_impl = ops->ctx;
```

Comment:
- [Debian ARM build fails](http://build.zfsonlinux.org/builders/Debian%208%20arm%20%28BUILD%29/builds/44239/steps/shell_2/logs/make)
- PPC and i386 have the same issue

```
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c: In function ‘sha256_alg_impl_benchmark’:
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:77:2: error: unknown type name ‘sha256_block_f’
  sha256_block_f sha256_impl = ops->ctx;
  ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:77:31: error: initialization makes integer from pointer without a cast [-Werror=int-conversion]
  sha256_block_f sha256_impl = ops->ctx;
                               ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:78:2: error: called object ‘sha256_impl’ is not a function or function pointer
  sha256_impl(sha2_ctx.state.s32, buffer, buffer_n >> 6);
  ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:77:17: note: declared here
  sha256_block_f sha256_impl = ops->ctx;
                 ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c: In function ‘sha512_alg_impl_benchmark’:
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:171:2: error: unknown type name ‘sha512_block_f’
  sha512_block_f sha512_impl = ops->ctx;
  ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:171:31: error: initialization makes integer from pointer without a cast [-Werror=int-conversion]
  sha512_block_f sha512_impl = ops->ctx;
                               ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:172:2: error: called object ‘sha512_impl’ is not a function or function pointer
  sha512_impl(sha2_ctx.state.s64, buffer, buffer_n >> 7);
  ^
/var/lib/buildbot/slaves/buildslave1/Debian_8_arm__BUILD_/build/zfs/module/icp/algs/sha2/sha2.c:171:17: note: declared here
  sha512_block_f sha512_impl = ops->ctx;
                 ^
```

Reply:
I mean, you put the definitions for sha{256,512}_block_f inside an `#if defined(__amd64) && defined(_KERNEL)`, so no wonder they're not turning up on not-amd64...

Reply:
Fixed :-)